pyg_lib/csrc/sampler/cpu/neighbor_kernel.cpp

```diff
@@ -475,53 +551,129 @@ sample(const std::vector<node_type>& node_types,
+              to_rel_type(k), pyg::utils::from_vector(edges_num_per_neighbors));
+        }
       }
     }
   });
 
   return std::make_tuple(out_row_dict, out_col_dict, out_node_id_dict,
-                         out_edge_id_dict);
+                         out_edge_id_dict, edges_num_dict, nodes_num_dict);
 }
 
 // Dispatcher //////////////////////////////////////////////////////////////////
 
-#define DISPATCH_SAMPLE(replace, directed, disjount, return_edge_id, ...) \
-  if (replace && directed && disjoint && return_edge_id)                  \
-    return sample<true, true, true, true>(__VA_ARGS__);                   \
-  if (replace && directed && disjoint && !return_edge_id)                 \
-    return sample<true, true, true, false>(__VA_ARGS__);                  \
-  if (replace && directed && !disjoint && return_edge_id)                 \
-    return sample<true, true, false, true>(__VA_ARGS__);                  \
-  if (replace && directed && !disjoint && !return_edge_id)                \
-    return sample<true, true, false, false>(__VA_ARGS__);                 \
-  if (replace && !directed && disjoint && return_edge_id)                 \
-    return sample<true, false, true, true>(__VA_ARGS__);                  \
-  if (replace && !directed && disjoint && !return_edge_id)                \
-    return sample<true, false, true, false>(__VA_ARGS__);                 \
-  if (replace && !directed && !disjoint && return_edge_id)                \
-    return sample<true, false, false, true>(__VA_ARGS__);                 \
-  if (replace && !directed && !disjoint && !return_edge_id)               \
-    return sample<true, false, false, false>(__VA_ARGS__);                \
-  if (!replace && directed && disjoint && return_edge_id)                 \
-    return sample<false, true, true, true>(__VA_ARGS__);                  \
-  if (!replace && directed && disjoint && !return_edge_id)                \
-    return sample<false, true, true, false>(__VA_ARGS__);                 \
-  if (!replace && directed && !disjoint && return_edge_id)                \
-    return sample<false, true, false, true>(__VA_ARGS__);                 \
-  if (!replace && directed && !disjoint && !return_edge_id)               \
-    return sample<false, true, false, false>(__VA_ARGS__);                \
-  if (!replace && !directed && disjoint && return_edge_id)                \
-    return sample<false, false, true, true>(__VA_ARGS__);                 \
-  if (!replace && !directed && disjoint && !return_edge_id)               \
-    return sample<false, false, true, false>(__VA_ARGS__);                \
-  if (!replace && !directed && !disjoint && return_edge_id)               \
-    return sample<false, false, false, true>(__VA_ARGS__);                \
-  if (!replace && !directed && !disjoint && !return_edge_id)              \
-    return sample<false, false, false, false>(__VA_ARGS__);
+#define DISPATCH_SAMPLE(replace, directed, disjount, return_edge_id, \
```

Comment:
Not a real expert on this one, but is there some way to auto-generate this? This gets pretty long :D

Reply:
Please look to the main reply.